tests/support/lib/random_buffer.c

```diff
@@ -7,65 +7,45 @@
 #include <stdint.h>
 #include <string.h>
 
+#include <sys/param.h>
 #include <sys/types.h>
 
 #include <hse/util/xrand.h>
 
 #include <hse/test/support/random_buffer.h>
 
 void
-randomize_buffer(void *buf, size_t len, unsigned int seed)
+randomize_buffer(char *buf, size_t len, unsigned int seed)
 {
-    unsigned int *tmp = (unsigned int *)buf;
-    uint last;
-    long int remain = len;
-    int i;
     struct xrand xr;
+    uint64_t rand_val;
 
     if (len == 0)
         return;
 
     xrand_init(&xr, seed);
-    for (i = 0; remain > 0; i++, remain -= sizeof(*tmp)) {
-        if (remain > sizeof(*tmp)) { /* likely */
-            tmp[i] = xrand64(&xr);
-        } else { /* unlikely */
-            last = xrand64(&xr);
-            memcpy(&tmp[i], &last, remain);
-        }
+    for (size_t remain = len; remain > 0; remain -= MIN(remain, sizeof(rand_val))) {
+        rand_val = xrand64(&xr);
+        memcpy(buf + len - remain, &rand_val, MIN(remain, sizeof(rand_val)));
     }
 }
 
 int
-validate_random_buffer(void *buf, size_t len, unsigned int seed)
+validate_random_buffer(char *buf, size_t len, unsigned int seed)
 {
-    unsigned int *tmp = (unsigned int *)buf;
-    unsigned int val;
-    char *expect = (char *)&val;
-    char *found;
-    long int remain = len;
-    int i;
     struct xrand xr;
+    uint64_t rand_val;
 
     if (len == 0)
         return -1; /* success... */
 
     xrand_init(&xr, seed);
-    for (i = 0; remain > 0; i++, remain -= sizeof(*tmp)) {
-        val = xrand64(&xr);
-        if ((remain >= sizeof(*tmp)) && (val != tmp[i])) { /* Likely */
-            return ((int)(len - remain));
-        } else if (remain < sizeof(*tmp)) { /* Unlikely */
-            found = (char *)&tmp[i];
-            if (memcmp(expect, found, remain)) {
-                /*
-                 * [HSE_REVISIT]
-                 * Miscompare offset might be off here
-                 */
-                return ((int)(len - remain));
-            }
-        }
+    for (size_t remain = len; remain > 0; remain -= MIN(remain, sizeof(rand_val))) {
+        rand_val = xrand64(&xr);
+        if (memcmp(&rand_val, buf + len - remain, MIN(remain, sizeof(rand_val))) != 0)
+            return (int)(len - remain);
     }
+
```

Comment:
that's prob going to be quite a bit slower now, couldn't you just have added code at the front to ensure correct alignment rather than rewrite everything?  at some point in the past these funcs were heavily used and very very slow...  maybe the compiler is smart enough to eliminate the memcpy in this case?  dunno...

Reply:
Let me play around with godbolt to check it out.

Reply:
Ok, after doing some testing, I don't think is an actual problem. This random_buffer stuff is only used a few times, with it being used the most in `cheap_test`. Looking at `cheap_test` on master and my branch on a debug build, the tests are all within the 4.5s-4s range.